src/pyroboplan/planning/path_shortcutting.py

```diff
@@ -0,0 +1,141 @@
+    Given a path and its corresponding length-normalized scalings between 0.0 and 1.0,
+    get the joint configuration at a specific scale value.
+
+    For example:
+      * 0.0 will return the start configuration of the path
+      * 1.0 will return the end configuration of the path
+      * 0.5 will return the configuration halfway along the length of the path
+
+    Parameters
+    ----------
+        q_path : list[array-like]
+            The list of joint configurations describing the path.
+        path_scalings : list[float]
+            The list of scaling values, between 0.0 and 1.0, at each of the path waypoints.
+        value : float
+            The value, between 0.0 and 1.0, at which to get the joint configuration along the path.
+
+    Return
+    ------
+        tuple (array-like, int)
+            A tuple containing the joint configuration at the specified scaling value along the path,
+            as well as the index corresponding to the next point along the path.
+    """
+    if value < 0 or value > 1:
+        raise ValueError("Scaling value must be in the range [0, 1]")
+
+    for idx in range(len(path_scalings)):
+        if value > path_scalings[idx]:
+            continue
+
+        delta_scale = (value - path_scalings[idx - 1]) / (
+            path_scalings[idx] - path_scalings[idx - 1]
+        )
+        return q_path[idx - 1] + delta_scale * (q_path[idx] - q_path[idx - 1]), idx
+
+
+def shortcut_path(
+    model, collision_model, q_path, max_iters=100, num_restarts=2, max_angle_step=0.05
+):
+    """
+    Performs path shortcutting by sampling two random points along the path and verifying
+    whether those two points can be connected directly without collisions.
+
+    This implementation is based on section 3.5.3 of:
+    https://motion.cs.illinois.edu/RoboticSystems/MotionPlanningHigherDimensions.html
+
+    Parameters
+    ----------
+        model : `pinocchio.Model`
+            The model of the robot.
+        collision_model : `pinocchio.Model`.
+            The model to use for collision checking.
+        q_path : list[array-like]
+            The list of joint configurations describing the path.
+        max_iters : int
+            The maximum iterations for randomly sampling shortcut points.
+        num_restarts : int
+            The number of restarts to try different random path shortcutting solutions.
+        max_angle_step : float
+            Maximum angle step, in radians, for collision checking along path segments.
+
+    Return
+    ------
+        list[array-like]
+            The shortest length path over all the shortcutting attempts.
+    """
+    best_path = q_path
+    best_path_length = get_path_length(best_path)
+
+    for _ in range(num_restarts + 1):
+        q_shortened = copy.deepcopy(q_path)
+        for _ in range(max_iters):
+            # Sample two points along the path length
+            path_scalings = get_normalized_path_scaling(q_shortened)
+            while True:
```

Comment:
note to self: this loop won't ever exit if the path has less than 3 points... so yeah, need to fix that.